MatrixSDK/Crypto/MXCrypto.m

```diff
@@ -1964,27 +1976,118 @@ - (MXHTTPOperation*)ensureOlmSessionsForDevices:(NSDictionary<NSString* /* userI
+    // The same but devices are listed by their identity key
+    NSMutableArray<NSString*> *devicesToClaim = [NSMutableArray array];
+    
+    // Devices (by their identity key) that are waiting for a response to /claim request
+    // That can be devices for which we are going to make a /claim request OR devices that
+    // already have a pending requests.
+    // Once we have emptied this array, we can call the success or the failure block. The
+    // operation is complete.
+    NSMutableArray<NSString*> *devicesInProgress = [NSMutableArray array];
+    
+    // Prepare the request for claiming one-time keys
     for (MXDeviceInfo *device in devicesWithoutSession)
     {
-        [usersDevicesToClaim setObject:oneTimeKeyAlgorithm forUser:device.userId andDevice:device.deviceId];
+        NSString *deviceIdentityKey = device.identityKey;
+        
+        // Claim only if a request is not yet pending
+        if (![ensureOlmSessionsInProgress containsObject:deviceIdentityKey])
+        {
+            [usersDevicesToClaim setObject:oneTimeKeyAlgorithm forUser:device.userId andDevice:device.deviceId];
+            [devicesToClaim addObject:deviceIdentityKey];
+            
+            [ensureOlmSessionsInProgress addObject:deviceIdentityKey];
+        }
+        
+        [devicesInProgress addObject:deviceIdentityKey];
     }
-
-    // TODO: this has a race condition - if we try to send another message
-    // while we are claiming a key, we will end up claiming two and setting up
-    // two sessions.
-    //
-    // That should eventually resolve itself, but it's poor form.
+    
+    NSLog(@"[MXCrypto] ensureOlmSessionsForDevices: %@ out of %@ sessions to claim one time keys", @(usersDevicesToClaim.count), @(devicesWithoutSession.count));
+    
+    
+    // Wait for the result of claim request(s)
+    // Listen to the dedicated notification
+    MXWeakify(self);
+    __block id observer;
+    observer = [[NSNotificationCenter defaultCenter] addObserverForName:kMXCryptoOneTimeKeyClaimCompleteNotification object:self queue:nil usingBlock:^(NSNotification * _Nonnull note) {
+        MXStrongifyAndReturnIfNil(self);
+        
+        NSMutableArray<NSString*> *devices = note.userInfo[kMXCryptoOneTimeKeyClaimCompleteNotificationDevicesKey];
+        NSError *error = note.userInfo[kMXCryptoOneTimeKeyClaimCompleteNotificationErrorKey];
+        
+        // Was it a /claim request for us?
+        if ([devicesInProgress mx_intersectArray:devices])
+        {
+            if (error)
+            {
+                NSLog(@"[MXCrypto] ensureOlmSessionsForDevices: Got a notification failure for %@ devices. Fail our current pool of %@ devices", @(devices.count), @(devicesInProgress.count));
+                
+                // Consider the failure for all requests of the current pool
+                [self->ensureOlmSessionsInProgress removeObjectsInArray:devicesInProgress];
```

Comment:
I think you are wrong here, you have to remove objects in `devices` array.

Indeed the error concerns only these devices. If I observe here an error about a device for which I did not trigger /claim request, your code will remove the devices for which I triggered a request. Let my request finish its job and perhaps work for other requests

Reply:
> I think you are wrong here, you have to remove objects in devices array.

I updated `devices`. It is clearly immutable. You cannot remove things from it.

> If I observe here an error about a device for which I did not trigger /claim request, your code will remove the devices for which I triggered a request.

We check `/ Was it a /claim request for us?` before going here. If the device is in `devicesInProgress`, a /claim request was triggered for it. The request could have been made by the current call of `ensureOlmSessionsForDevices` or a previous one. We do not take care and we do not need to.

> Let my request finish its job and perhaps work for other requests

We fail for all devices in `devicesInProgress ` because we cannot have a grey answer or half success.
If `ensureOlmSessionsForDevices` is called 2 times in parallel with `[DeviceA]` then `[DeviceA, DeviceB, DeviceC]`. If /claim for DeviceA fails, we need to fail both calls even if /claim worked for DeviceB & DeviceC.